host/sgx/create.c

```diff
@@ -589,57 +624,85 @@ oe_result_t oe_sgx_validate_enclave_properties(
 }
 
 #ifdef OE_WITH_EXPERIMENTAL_EEID
-static oe_result_t _add_eeid_marker_page(
+static oe_result_t _save_eeid_info(
     oe_sgx_load_context_t* context,
-    oe_enclave_t* enclave,
-    size_t image_size,
     uint64_t entry_point,
     oe_sgx_enclave_properties_t* props,
     uint64_t* vaddr)
 {
-    oe_result_t result = OE_UNEXPECTED;
-    oe_eeid_t* eeid = context->eeid;
+    const oe_enclave_setting_eeid_t* eeid_setting = context->eeid_setting;
 
-    if (eeid && is_eeid_base_image(props) &&
-        context->type == OE_SGX_LOAD_TYPE_CREATE)
+    if (eeid_setting)
     {
-        // Finalize the memory settings
-        props->header.size_settings = eeid->size_settings;
+        /* EEID must fit on a single page. */
+        size_t signature_size = sizeof(sgx_sigstruct_t);
+        size_t eeid_size = sizeof(oe_eeid_t) + signature_size;
+        if (eeid_size > OE_PAGE_SIZE)
+            return OE_INVALID_PARAMETER;
+
+        oe_eeid_t* eeid = NULL;
+        oe_create_eeid_sgx(&eeid);
+
+        /* Finalize the memory settings */
+        if (context->type == OE_SGX_LOAD_TYPE_CREATE &&
+            is_eeid_base_image(&props->header.size_settings))
+            props->header.size_settings = eeid_setting->size_settings;
+        else
+            eeid->size_settings = props->header.size_settings;
 
-        // Record EEID information
+        /* Record EEID information */
         eeid->version = OE_EEID_VERSION;
         oe_sha256_context_t* hctx = &context->hash_context;
         oe_sha256_save(hctx, eeid->hash_state.H, eeid->hash_state.N);
+        eeid->size_settings = props->header.size_settings;
         eeid->entry_point = entry_point;
         eeid->vaddr = *vaddr;
-        eeid->signature_size = sizeof(sgx_sigstruct_t);
-        memcpy(
-            eeid->data + eeid->data_size,
-            (uint8_t*)&props->sigstruct,
-            sizeof(sgx_sigstruct_t));
+        eeid->signature_size = signature_size;
+        memcpy(eeid->signature, props->sigstruct, signature_size);
+
+        /* Compute config hash (all zero during signing) */
+        memset(eeid->config_id, 0, sizeof(eeid->config_id));
```

Comment:
This implementation locks the definition of `config_id`, which should be applicaiton specific. A default definition of `conif_di`=SHA256(`eeid_setting.data`) is OK, but we should allow the enclave developer to override it. Including `config_id` (and `config_svn`) in `eeid_setting` would allow the developer to choose whether to use the default definition of `config_id`. If the developer sets `eeid_setting.data_size` as non zero, the default defintion is choosen. If the developer sets `eeid_setting.data_size`=0, the implementation should load `config_id`from `eeid_setting.config_id`as it is. The enclave developer can implement an explicit ECALL to add additional content post enclave initialization, and fully control the expected relationship between `config_id` (and `config_svn`) and the additional content.